File: src/test/java/com/ludo/study/studymatchingplatform/study/fixture/StudyFixture.java
package com.ludo.study.studymatchingplatform.study.fixture;

import java.time.LocalDateTime;

import com.ludo.study.studymatchingplatform.study.domain.Category;
import com.ludo.study.studymatchingplatform.study.domain.Platform;
import com.ludo.study.studymatchingplatform.study.domain.Study;
import com.ludo.study.studymatchingplatform.study.domain.StudyStatus;
import com.ludo.study.studymatchingplatform.study.domain.Way;
import com.ludo.study.studymatchingplatform.user.domain.User;

public class StudyFixture {

	public static Study createStudy(StudyStatus studyStatus, String title, Way way, Category category, User user,
									int participantCount, int participantLimit, Platform platform
	) {
		return Study.builder()
				.status(studyStatus)
				.platform(platform)
				.category(category)
				.owner(user)
				.title(title)
				.way(way)
				.participantCount(participantCount)
				.participantLimit(participantLimit)
				.startDateTime(LocalDateTime.now())
				.endDateTime(LocalDateTime.now())
				.build();
	}

	public static Study createStudy(String title, Category category, User user,
									int participantLimit, Platform platform
	) {
		return createStudy(
				StudyStatus.RECRUITING,
				title,
				Way.ONLINE,
				category,
				user,
				0,
				participantLimit,
				platform
		);
	}

}

File: src/e2e/fixtures/study-fixture.ts
import { addDays } from "date-fns/addDays";
import { localNow } from "../helpers/datetime-helper";
import { CreateStudyRequest, UpdateStudyRequest } from "../types/study-types";
import { randCategoryId } from "./category-fixture";
import { randOneOf, randPositionId, randUrl } from "./position-fixture";
import { randInt, randStr } from "./rand-fixture";

export type FakeCreateStudyRequestArgs = Partial<CreateStudyRequest>;
export type FakeUpdateStudyRequestArgs = FakeCreateStudyRequestArgs;

export function fakeCreateStudyRequest({
  categoryId,
  endDateTime,
  participantLimit,
  platform,
  platformUrl,
  positionId,
  startDateTime,
  title,
  way,
}: FakeCreateStudyRequestArgs = {}): CreateStudyRequest {
  const now = localNow();
  return {
    title: title ?? randStr(10, 20),
    categoryId: categoryId ?? randCategoryId(),
    positionId: positionId ?? randPositionId(),
    way: way ?? randOneOf("ONLINE", "OFFLINE"),
    platform: platform ?? randOneOf("GATHER", "GOOGLE_MEET"),
    platformUrl: platformUrl ?? randUrl(),
    participantLimit: participantLimit ?? randInt(4, 8),
    startDateTime: startDateTime ?? now,
    endDateTime: endDateTime ?? addDays(now, 10).toISOString(),
  };
}

export function fakeUpdateStudyRequest({
  categoryId,
  endDateTime,
  participantLimit,
  platform,
  platformUrl,
  positionId,
  startDateTime,
  title,
  way,
}: FakeUpdateStudyRequestArgs = {}): UpdateStudyRequest {
  const now = localNow();
  return {
    title: title ?? randStr(10, 20),
    categoryId: categoryId ?? randCategoryId(),
    positionId: positionId ?? randPositionId(),
    way: way ?? randOneOf("ONLINE", "OFFLINE"),
    platform: platform ?? randOneOf("GATHER", "GOOGLE_MEET"),
    platformUrl: platformUrl ?? randUrl(),
    participantLimit: participantLimit ?? randInt(4, 8),
    startDateTime: startDateTime ?? now,
    endDateTime: endDateTime ?? addDays(now, 10).toISOString(),
  };
}
